### PR TITLE
feat: meaningful error pages

### LINKED ISSUE #133: Return meaningful errors
iiuc, right now, we seem to return error that produces default error from browser engine:

https://github.com/ipfs-shipyard/helia-service-worker-gateway/blob/41d2cffcff5bc3061212fe13aadb5df9c1a3f64c/src/sw.ts#L389

https://github.com/ipfs-shipyard/helia-service-worker-gateway/blob/41d2cffcff5bc3061212fe13aadb5df9c1a3f64c/src/sw.ts#L387

This makes it hard for user to reason what went wrong, and near to impossible to fill a meaningful error report.

## Example: failure to fetch block during path walk

If you walk https://inbrowser.dev/ipns/en.wikipedia-on-ipfs.org long enough, you may hit a hiccup:

> 2024-04-11-171819_658x702_scrot

### Proposed improvement

If we encounter an error while generating a Response within a service worker, we should always catch the error and respond with a meaningful HTML error page. 

- [ ] return `text/html`
- [ ] human-readable
- [ ] include error details (short message, full JS error's stack)
- [ ] do this for [timeouts](https://github.com/ipfs-shipyard/helia-service-worker-gateway/blob/41d2cffcff5bc3061212fe13aadb5df9c1a3f64c/src/sw.ts#L387) and redirects as well
- [ ] **VERY IMPORTANT UX:** have "golden path" for fixing hiccups. HTML big green button "Retry" that allows user to reload page and try again 

### Pseudocode

```js
async function handleRequest(request) {
  try {
    const response = await fetch(request)
    return response
  } catch (error) {
    // If an error occurs, respond with a custom error page including error details
    const errorResponse = new Response(`
      <h1>Oops! Something went wrong inside of Service Worker IPFS Gateway.</h1>
      <p><button onclick="window.location.reload(true);">🔃 Click here to retry</button>.</p>
      <p>Error details:</p>
      <p>${error.message}</p>
      <pre>${error.stack}</pre>
    `, {
      status: 500,
      statusText: 'Internal Service Worker Gateway Error',
      headers: new Headers({
        'Content-Type': 'text/html'
      })
    })
    return errorResponse
  }
}
```

### PR DESCRIPTION
**Pull request by SgtPooki**
- feat: basic error page handling on errors
- feat: sw info on error page

## Title

feat: meaningful error pages

## Description

Fixes https://github.com/ipfs-shipyard/service-worker-gateway/issues/133

## Notes & open questions

We need to provide more stack trace information from `@helia/verified-fetch` to make this truly useful.

I've been debugging this at http://en-wikipedia--on--ipfs-org.ipns.localhost:3000/wiki/Oak/foobar 

With the changes in this PR so far, the error page shows:

image

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation if necessary (this includes comments as well)
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: src/sw.ts [hunk 1]
```diff
@@ -476,6 +480,83 @@ async function fetchHandler ({ path, request, event }: FetchHandlerArg): Promise
...
+  let text: string | null = null
+  if (responseContentType != null) {
+    if (responseContentType.includes('text/html')) {
+      return fetchResponse
+    } else if (responseContentType.includes('application/json')) {
+      // we may eventually provide error messaging from @helia/verified-fetch
+      json = await fetchResponse.json()
+    } else {
+      text = await fetchResponse.text()
+      json = { error: { message: fetchResponse.statusText, stack: text } }
```
Comment (lidel) [inline src/sw.ts]: nit: for this `text` path, is the plan to fix low-hanging fruit for now and include `err` in message returned in [`return badGatewayResponse(resource, 'Error walking path')`](https://github.com/ipfs/helia-verified-fetch/blob/089635d6cd5b10aefbed013e95637ddb90b166e5/packages/verified-fetch/src/verified-fetch.ts#L257) for now, making it more useful: `Bad Gateway: Error walking path: reason`

If so, we should not pretend it is stack, but concatenate error message:

```suggestion
      text = await fetchResponse.text()
      json = { error: { message: `${fetchResponse.statusText}: ${text}`, stack: null } }
```

File: src/sw.ts [hunk 2]
```diff
@@ -476,6 +480,83 @@ async function fetchHandler ({ path, request, event }: FetchHandlerArg): Promise
...
+ * TODO: more service worker details
+ */
+async function getServiceWorkerDetails (): Promise<ServiceWorkerDetails> {
+  const registration = self.registration
+  const state = registration.installing?.state ?? registration.waiting?.state ?? registration.active?.state ?? 'unknown'
+
+  return {
+    config: await getConfig(),
+    crossOriginIsolated: self.crossOriginIsolated,
+    installTime: (new Date(firstInstallTime)).toISOString(),
```
Comment (lidel) [inline src/sw.ts]: nit: UTC makes debugging easier
```suggestion
    installTime: (new Date(firstInstallTime)).toUTCString(),
```

**Comment by SgtPooki** [comment]
Error page with latest updates is:

image

**Comment by SgtPooki** [comment]
Latest page visual with [`06eab49` (#195)](https://github.com/ipfs-shipyard/service-worker-gateway/pull/195/commits/06eab4921ab5e8df913ccba89d42756e93d0f04c) is:

image

**Pull request merged**